[main] Update documentation for kubelet drop-in config dir (#44985)
kubernetes/website: content/en/docs/tasks/administer-cluster/kubelet-config-file.md

Reviewer: Overall, there's a lot of detail here that belongs in a reference page.

Someone trying to follow the existing task page will find that task more confusing if we make these changes. Can we move the details elsewhere? For example, within https://kubernetes.io/docs/concepts/cluster-administration/?

defaults for the `KubeletConfiguration` version apply.
In the above example, this version is `kubelet.config.k8s.io/v1beta1`.

> ## Viewing the kubelet configuration

Reviewer 2: what I am looking for when asking for /configz documentation is a clear statement on what that config represents:

1. Does it merge-in all the command line arguments into it?
2. Does it merge-in all the drop-in configuration files?
3.  Does it populate all the defaults?

In the example in this section, the feature gate consist of a single `"AllAlpha": false`. I wonder if it was configured by you explicitly or this is what it returns out of the box with nothing pre-configured?

Author: >what I am looking for when asking for /configz documentation is a clear statement on what that config represents:

>    Does it merge-in all the command line arguments into it?
>   Does it merge-in all the drop-in configuration files?
>  Does it populate all the defaults?

Do you want me to include this point under the Drop-in directory for kubelet configuration files? Otherwise, it may not make much sense without that background knowledge

Author: >In the example in this section, the feature gate consist of a single "AllAlpha": false. I wonder if it was configured by you explicitly or this is what it returns out of the box with nothing pre-configured?

I didn't configure the Kubelet config; it came out of the box.

Reviewer 2: > > what I am looking for when asking for /configz documentation is a clear statement on what that config represents:
> 
> > Does it merge-in all the command line arguments into it?
> > Does it merge-in all the drop-in configuration files?
> > Does it populate all the defaults?
> 
> Do you want me to include this point under the Drop-in directory for kubelet configuration files? Otherwise, it may not make much sense without that background knowledge

it does make sense when we talk about arguments and such. If it were separate page about /configz, than it will definitely make sense and we can simply refer to it. If we don't have it as of now - simply mentiuoning that this is a final merged config will make the most sense.

To inspect the kubelet configuration, follow these steps:

> 1. Start a proxy server using `kubectl proxy` in your terminal.

Reviewer 3: Is this code always going to work? 

I don't know much about `kubectl proxy` but this seems that it would be locked down from most users using managed cloud services?

And do we need to pass any certs or tls for curl for this to work?

Author: There's another way to query the kubelet config
```bash
$ kubectl get --raw "/api/v1/nodes/<node-name>/proxy/configz" | jq
```
which doesn't involve passing any certs,

Author: Yeah, that's tricky. I wrote this documentation with the normal use case in mind.

Reviewer 3: I don't know what normal would be in this case. I would usually suggest mentioning links to `kubectl proxy` or something more detailed in case it doesn't work for the user..

```bash
$ kubectl proxy
Starting to serve on 127.0.0.1:8001

```
2. Open another terminal window and use `curl` to fetch the kubelet configuration. 
Replace `<node-name>` with the actual name of your node:

```bash
$ curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .

> {

Author: Let me know if displaying the entire kubelet config makes sense here.

    "enableControllerAttachDetach": true,
    "makeIPTablesUtilChains": true,
    "iptablesMasqueradeBit": 14,
    "iptablesDropBit": 15,
    "featureGates": {
      "AllAlpha": false
    },
    "failSwapOn": false,
    "memorySwap": {},
    "containerLogMaxSize": "10Mi",
    "containerLogMaxFiles": 5,
    "configMapAndSecretChangeDetectionStrategy": "Watch",
    "enforceNodeAllocatable": [
      "pods"
    ],
    "volumePluginDir": "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/",
    "logging": {
      "format": "text",
      "flushFrequency": "5s",
      "verbosity": 3,
      "options": {
        "json": {
          "infoBufferSize": "0"
        }
      }
    },
    "enableSystemLogHandler": true,
    "enableSystemLogQuery": false,
    "shutdownGracePeriod": "0s",
    "shutdownGracePeriodCriticalPods": "0s",
    "enableProfilingHandler": true,
    "enableDebugFlagsHandler": true,
    "seccompDefault": false,
    "memoryThrottlingFactor": 0.9,
    "registerNode": true,
    "localStorageCapacityIsolation": true,
    "containerRuntimeEndpoint": "unix:///var/run/crio/crio.sock"
  }
}
```

## Drop-in directory for kubelet configuration files {#kubelet-conf-d}

> {{<feature-state for_k8s_version="v1.28" state="alpha" >}}

Author: https://github.com/kubernetes/website/pull/44985#discussion_r1480073608

> {{<feature-state for_k8s_version="v1.28" state="alpha" >}}

Author: This change is specific to Alpha and will be merged into the main branch. Subsequently, I will create another PR targeting the `1.30-dev` branch for transitioning to beta.

> {{<feature-state for_k8s_version="v1.28" state="alpha" >}}

Reviewer 3: Ah okay. Missed that.

```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
authorization:
  mode: AlwaysAllow
  webhook:
    cacheAuthorizedTTL: "8m"
    cacheUnauthorizedTTL: "45s"
address: "192.168.0.8"
```

#### Lists/Slices
You can overidde the slices/lists values of the kubelet configuration, however, the entire list gets overridden during the merging process. For example, you can override the `clusterDNS` list as follows:

Main kubelet configuration file contents:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
clusterDNS:
  - "192.168.0.9"
  - "192.168.0.8"
```

Contents of a file in `--config-dir` directory:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
clusterDNS:
  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"
```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
clusterDNS:
  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"
```

#### Maps, including Nested Structures

> Individual fields in maps, regardless of their value types (boolean, string, etc.), can be selectively overridden. However, for `map[string][]string`, the entire list associated with a specific field gets overridden. Let's understand this better with an example, particularly on fields like `featureGates` and `staticPodURLHeader`:

Author: ok, so the kubelet config by default sets the value for `AllAlpha` to false. It should be there in the final merged config. Let me fix this part.

Main kubelet configuration file contents:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
featureGates:
  MemoryQoS: true
staticPodURLHeader:
  kubelet-api-support:
  - "Authorization: 234APSDFA"
  - "X-Custom-Header: 123"
  custom-static-pod:
  - "Authorization: 223EWRWER"
  - "X-Custom-Header: 456"
```

Contents of a file in `--config-dir` directory:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
featureGates:
  MemoryQoS: false
  KubeletTracing: true
staticPodURLHeader:
  kubelet-api-support:

>   - "Authorization: 8945AFSG1"

Reviewer 4: Can you show an example of dropping or not specififying a field in the override file?

Author: Fixed

    "shutdownGracePeriodCriticalPods": "0s",
    "enableProfilingHandler": true,
    "enableDebugFlagsHandler": true,
    "seccompDefault": false,
    "memoryThrottlingFactor": 0.9,
    "registerNode": true,
    "localStorageCapacityIsolation": true,
    "containerRuntimeEndpoint": "unix:///var/run/crio/crio.sock"
  }
}
```

## Drop-in directory for kubelet configuration files {#kubelet-conf-d}

{{<feature-state for_k8s_version="v1.28" state="alpha" >}}

You can specify a drop-in configuration directory for the kubelet. By default, the kubelet does not look
for drop in configuration files anywhere; you must specify a path.
For example: `--config-dir=/etc/kubernetes/kubelet.conf.d`

For Kubernetes v1.28 to v{{< skew currentVersion >}}, you  can only specify `--config-dir` if you also set
the environment variable `KUBELET_CONFIG_DROPIN_DIR_ALPHA` for the kubelet process (the value
of that variable does not matter).

You can use the kubelet configuration directory in a similar way to the kubelet configuration file.

{{< note >}}
The suffix of a valid kubelet drop-in configuration file must be `.conf`. For instance: `99-kubelet-address.conf`
{{< /note >}}

> The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. This offers you flexibility in how you manage and combine kubelet configuration that comes from different sources. However, it's important to note that the behavior varies based on the data type of the configuration fields. Here are some examples of how different data types behave during configuration merging:

Reviewer 2: ```suggestion
The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. Files may contain partial configuration and may not be a valid config files by themselves, validation only performed on a final resulting configuration file. This offers you flexibility in how you manage and combine kubelet configuration that comes from different sources. However, it's important to note that the behavior varies based on the data type of the configuration fields. Here are some examples of how different data types behave during configuration merging:
```

## Drop-in directory for kubelet configuration files {#kubelet-conf-d}

> 

Reviewer: ```suggestion

{{<feature-state for_k8s_version="v1.28" state="alpha" >}}

```

(To mark when the feature went alpha)

> In Kubernetes v1.30.0, the drop-in configuration directory feature has reached Beta status. This allows the kubelet to support a drop-in configuration directory, which can be specified using the `--config-dir` flag. By default, it is set to `""`, indicating that the feature is disabled by default.

Reviewer 5: I think this should be 1.29.0 and update with the new correct behavior. We can update to beta once we actually update the code to beta

Author: Fixed

> In Kubernetes v1.29.0, the drop-in configuration directory feature allows the kubelet to support a drop-in configuration directory, which can be specified using the `--config-dir` flag. By default, it is set to `""`, indicating that the feature is disabled by default. Users need to expliclty opt-in for this feature by setting the `--config-dir` flag to a valid directory path.

Reviewer: ```suggestion
You can specify a drop-in configuration directory for the kubelet. By default, the kubelet does not look
for drop in configuration files anywhere; you must specify a path.
For example: `--config-dir=/etc/kubernetes/kubelet.conf.d`
```

> 

Reviewer: ```suggestion

For Kubernetes v1.28 to v{{< skew currentVersion >}}, you  can only specify `--config-dir` if you also set
the environment variable `KUBELET_CONFIG_DROPIN_DIR_ALPHA` for the kubelet process (the value
of that variable does not matter).
```

This change is appropriate for a PR targeting _main_; it will render right. Then, once _this_ is merged to main and has been onwards merged to _dev-1.30_, update that branch to mark the feature as beta. You would drop the
`v1.28 to v{{< skew currentVersion >}}` text and code it as (eg) `v1.28 and v1.29`.

You cannot specify the drop-in configuration directory using the kubelet configuration file; only the CLI argument `--config-dir` can set it.

> One can use the kubelet configuration directory in a similar way to the kubelet config file.

Reviewer: (nit / cleanup)
```suggestion
You can use the kubelet configuration directory in a similar way to the kubelet configuration file.
```

## Drop-in directory for kubelet configuration files {#kubelet-conf-d}

{{<feature-state for_k8s_version="v1.28" state="alpha" >}}

You can specify a drop-in configuration directory for the kubelet. By default, the kubelet does not look
for drop in configuration files anywhere; you must specify a path.
For example: `--config-dir=/etc/kubernetes/kubelet.conf.d`

> For Kubernetes v1.28 to v{{< skew currentVersion >}}, you  can only specify `--config-dir` if you also set

Reviewer 5: wouldn't this be `v1.28 to v1.29` (instead of currentVersion, which would be 1.30)

Author: I followed [this](https://github.com/kubernetes/website/pull/44985#discussion_r1475839135) comment

Reviewer: The current version of Kubernetes is v1.29. 

> I've suggested changes that I think are appropriate for a PR targeting _main_; it will render right. Then, once this is merged to _main_ and has been onwards merged to _dev-1.30_, send in a second PR to promote the feature to beta. The second PR should have _dev-1.30_ as its base branch.

The approach I've suggest would work even if the k/k update to kubelet drop-in config didn't land until a later minor release.

Author: @sftim does the latest push make sense?

Reviewer: Yep

the environment variable `KUBELET_CONFIG_DROPIN_DIR_ALPHA` for the kubelet process (the value
of that variable does not matter).

> You can use the kubelet configuration directory in a similar way to the kubelet configuration file.

Reviewer: nit: I'd put a whilespace line after this one.

Author: Fixed

> ## Drop-in directory for kubelet configuration files {#kubelet-conf-d}

Reviewer 2: Beta graduation criteria ask to document the /configz endpoint as this is a recommended and only way for a cluster admin to review the resuting configuration file. Can you please send a PR here for this?

Link to graduation criteria: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3983-drop-in-configuration/README.md#beta

Author: Sure, does it make sense to add that before the drop-in directory point with the title `Viewing the kubelet configuration`?

Reviewer 2: this is the first mention of `/configz` in docs ever. I think it deserves it's own page. @sftim what would be the best place to document the `/configz` endopint?

Reviewer: Hi. Best place would be within https://kubernetes.io/docs/reference/node/

the environment variable `KUBELET_CONFIG_DROPIN_DIR_ALPHA` for the kubelet process (the value
of that variable does not matter).

You can use the kubelet configuration directory in a similar way to the kubelet configuration file.

{{< note >}}
The suffix of a valid kubelet drop-in configuration file must be `.conf`. For instance: `99-kubelet-address.conf`
{{< /note >}}

The kubelet processes files in its config drop-in directory by sorting the **entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration files with a later sort order can override earlier files. These files may contain partial configurations and might not be valid config files by themselves; validation is only performed on the final resulting configuration file.  This offers you flexibility in how you manage and combine kubelet configuration that comes from different sources. However, it's important to note that the behavior varies based on the data type of the configuration fields. Here are some examples of how different data types behave during configuration merging:

> #### Singular/Embedded struct fields

Reviewer: ```suggestion
#### Configuration merge behavior for simple values
```

> The merging process effectively handles the overriding of singular or embedded struct fields in kubelet configuration. For instance, you may want a baseline kubelet configuration for all nodes, but you may want to customize the `address` and `authorization` fields. This can be done as follows:

Reviewer: Readers likely have no idea what you mean by “embedded struct” (and we should not force them to know).

Main kubelet configuration file contents:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
authorization:
  mode: Webhook
  webhook:
    cacheAuthorizedTTL: "5m"
    cacheUnauthorizedTTL: "30s"
serializeImagePulls: false
address: "192.168.0.1"
```

Contents of a file in `--config-dir` directory:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
authorization:
  mode: AlwaysAllow
  webhook:
    cacheAuthorizedTTL: "8m"
    cacheUnauthorizedTTL: "45s"
address: "192.168.0.8"
```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
authorization:
  mode: AlwaysAllow
  webhook:
    cacheAuthorizedTTL: "8m"
    cacheUnauthorizedTTL: "45s"
address: "192.168.0.8"
```

> #### Lists/Slices

Reviewer: (nit)
```suggestion
#### Configuration merge behavior for lists and maps
``

> You can overidde the slices/lists values of the kubelet configuration, however, the entire list gets overridden during the merging process. For example, you can override the `clusterDNS` list as follows:

Reviewer: I wouldn't use “slice” here; YAML calls it a _mapping_.

  KubeletTracing: true
  DynamicResourceAllocation: true
staticPodURLHeader:
  kubelet-api-support:
  - "Authorization: 234APSDFA"
  - "X-Custom-Header: 123"
  custom-static-pod:
  - "Authorization: 223EWRWER"
  - "X-Custom-Header: 345"
```

### Kubelet configuration merging order

On startup, the kubelet merges configuration from:

* Feature gates specified over the command line (lowest precedence).
* The kubelet configuration.
* Drop-in configuration files, according to sort order.
* Command line arguments exclude feature gates (highest precedence).

{{< note >}}
The config drop-in dir mechanism for the kubelet is similar but different from how the `kubeadm` tool allows you to patch configuration. The `kubeadm` tool uses a specific [patching strategy](/docs/setup/production-environment/tools/kubeadm/control-plane-flags/#patches) for its configuration, where configuration files are merged based on an optional file name suffix. The kubelet determines the order of merges based on sorting the **suffixes** alphanumerically.
{{< /note >}}

This produces the same outcome as if you used the [single configuration file](#create-the-config-file) used in the earlier example.

> ## Viewing the kubelet configuration

Reviewer: I would actually promote this to be its own task page (perhaps also its own commit, or own PR).

> To inspect the kubelet configuration, follow these steps:

Reviewer: Do you need any access grants to be able to do this?

> 1. Start a proxy server using [`kubectl proxy`](/docs/reference/generated/kubectl/kubectl-commands#proxy) in your terminal.

Reviewer 6: ```suggestion
1. Start a proxy server using [`kubectl proxy`](/docs/reference/kubectl/generated/kubectl_proxy/) in your terminal.
```

```bash

> $ kubectl proxy

Reviewer 6: We [do not include the command prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt), so please remove it.

Starting to serve on 127.0.0.1:8001

```
2. Open another terminal window and use `curl` to fetch the kubelet configuration. 
Replace `<node-name>` with the actual name of your node:

```bash

> curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .

Reviewer: If requiring `curl` and `jq`, make sure to list these as task prerequisites.

> {

Reviewer: We aim to [separate commands from output](https://kubernetes.io/docs/contribute/style/style-guide/#separate-commands-from-output).

> $ curl -X GET http://127.0.0.1:8001/api/v1/nodes/<node-name>/proxy/configz | jq .

Reviewer 6: Same here, please remove the prompt.

Main kubelet configuration file contents:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
clusterDNS:
  - "192.168.0.9"
  - "192.168.0.8"
featureGates:
  MemoryQoS: true
```

Contents of a file in `--config-dir` directory:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
clusterDNS:
  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"
featureGates:
  KubeletTracing: false
```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
clusterDNS:
  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"
featureGates:
  MemoryQoS: true
  KubeletTracing: false
```

On startup, the kubelet merges configuration from:

* Feature gates specified over the command line (lowest precedence).
* The kubelet configuration.
* Drop-in configuration files, according to sort order.
* Command line arguments (highest precedence).

{{< note >}}

> Kubeadm utilizes a specific [patching strategy](/docs/setup/production-environment/tools/kubeadm/control-plane-flags.md/#patches), where configuration files are merged based on an optional file name suffix, determining the order in an alphanumeric manner. Conversely, the kubelet config drop-in directory feature processes these configuration files in lexical order based on file names, i.e., configurations in files with lower numeric prefixes are merged first, followed by higher numeric prefixes. This offers users flexibility in managing Kubelet configurations across multiple nodes.

Reviewer: How about:
```suggestion
The kubelet processes files in its config drop-in directory using a similar mechanism, sorting the
**entire file name** alphanumerically. Files with that sort earlier are processed first, and configuration
files with a later sort order can override earlier files.

This offers you flexibility in how you manage and combine kubelet configuration that comes from
different sources.

---

The config drop in mechanism for the kubelet is similar but different from how the `kubeadm`
tool allows you to patch configuration. The `kubeadm` tool uses a specific
[patching strategy](/docs/setup/production-environment/tools/kubeadm/control-plane-flags/#patches)
for its configuration, where configuration files are merged based on an optional file name suffix.
The kubelet determines the order of merges based on sorting the **suffixes** alphanumerically.
```

Reviewer: (this might be long enough to merit being an item with its own heading, rather than a note)

Author: Let me know if the latest push makes sense here.

Author: @sftim could you PTAL?

address: "192.168.0.8"
```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
address: "192.168.0.8"
```

> Additionally, you can change the slices/lists values of the kubelet configuration. For example, you can override the `clusterDNS` and `featureGates` fields as follows:

Reviewer 2: please give an example with overriding values

Reviewer 2: example: `failSwapOn ` - disabled explicitly in main config and enabled in child config. May be a common scenario to enable swap via drop-in config

Author: This is covered in the first example.

Main kubelet configuration file contents:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
clusterDNS:
  - "192.168.0.9"
  - "192.168.0.8"
featureGates:
  MemoryQoS: true
```

Contents of a file in `--config-dir` directory:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration

> clusterDNS:

Reviewer 2: how eviction collections will be merged? Will the whole structure be overriden or one value added?

Reviewer 2: Same question about logging - will the whole structure be overriden or individual fields?

Author: The merging logic is targeted for the individual fields. If the intention is to override fields, then mention only those fields in the drop-in configs.

Reviewer 2: individual high level fields or all of them? I simply don't know how it works and this documentation doesn't give a clear understanding.

Reviewer 2: evictionHard and evictionSoft is coming from - what if I want to configur MemorySwap. I need to enable feature gate, and potentially want to disable eviction hard and evictionSoft. Is it possible to do with a single drop-in file?

Author: >evictionHard and evictionSoft is coming from - what if I want to configur MemorySwap. I need to enable feature gate, and potentially want to disable eviction hard and evictionSoft. Is it possible to do with a single drop-in file?

Yes, you can set the values in a drop-in config for this use case.

Author: >individual high level fields or all of them? I simply don't know how it works and this documentation doesn't give a clear understanding.

Please scratch the previous response.

>how eviction collections will be merged? Will the whole structure be overriden or one value added?

In this case, the entire structure will be overridden. If you want to change just one field, you can certainly do that. For example, in the case of `evictionHard`, you can alter the `memory.available` to a new value. However, you have to set values for other fields, perhaps using default values or any other values you desire, to make that happen.

>   - "192.168.0.2"

Reviewer 2: is there a way to clean up the list? Is it a typical requirement?

Author: Not sure if I understood your question. Let me take a shot. When you refer to the main config, you are talking about the base kubelet config. The intention behind adding drop-in files is to modify or clean up the values in the base configuration.

>   - "192.168.0.3"

Reviewer 2: can the previously defined single value be removed from the list?

Author: Here's the caveat: you cannot remove a single value from the list. You need to replace the entire list with the new values. If you want to maintain the original values as well, you need to mention that in the new drop-in config.

  - "192.168.0.5"
featureGates:

>   KubeletTracing: false

Reviewer 2: can the feature gate be overriden? Will there be an error or we will follow the priority order to decide on the final value?

```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
clusterDNS:
  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"
featureGates:
  MemoryQoS: true
  KubeletTracing: false
```

On startup, the kubelet merges configuration from:

* Feature gates specified over the command line (lowest precedence).
* The kubelet configuration.
* Drop-in configuration files, according to sort order.

> * Command line arguments (highest precedence).

Reviewer 2: it's hard to understand how feature gates from the command line has lowest precedence and command line - highest? Maybe say here that this excludes feature gates

Author: Sure, I will add that information.

address: "192.168.0.8"
```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
address: "192.168.0.8"
```
Additionally, you can overidde the slices/lists values of the kubelet configuration. For example, you can override the `clusterDNS` and `featureGates` fields as follows:

Main kubelet configuration file contents:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"
clusterDNS:
  - "192.168.0.9"
  - "192.168.0.8"
featureGates:
  MemoryQoS: true
```

Contents of a file in `--config-dir` directory:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
clusterDNS:
  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"
featureGates:
  MemoryQoS: false
  KubeletTracing: true
```

The resulting configuration will be as follows:
```yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
port: 20250
serializeImagePulls: false
evictionHard:
  memory.available: "200Mi"

> clusterDNS:

Reviewer 2: My understanding is that all lists are being overriden as a whole. Not merged. If so - will it be the problem with this specific field? Will one ever want to simply add a new DNS via the drop-in config?

Author: Patching is not currently supported in the existing implementation. Therefore, while adding a new DNS value is possible, it results in the removal of previous values in the `clusterDNS` list.

  - "192.168.0.2"
  - "192.168.0.3"
  - "192.168.0.5"

> featureGates:

Reviewer 2: will the speciall `all/alpha=TRUE` work in drop-in config? Will it respect previously disabled alphas or will force enable them?

Author: >will the speciall all/alpha=TRUE work in drop-in config

It should work

>Will it respect previously disabled alphas or will force enable them?

The latter

Reviewer 4: @sohankunkerkar can you include this in the website example?

Author: >Will it respect previously disabled alphas or will force enable them?

The above logic will be handled by the Kubelet. This functionality aims to replace the corresponding field value.

>   MemoryQoS: false

Reviewer 2: if you didn't specify MemoryQoS in the drop in config - will it stay true or will change to whatever is default in k8s?

Author: If you don't mention the `MemoryQoS` field in the drop-in config, it will display the value mentioned in the base kubelet config unless you set the value via a command-line flag, which takes higher precedence over the drop-in config values.
